File: src/main/kotlin/ting/java/leetcode/hashtable/JewelsAndStones.java
package ting.java.leetcode.hashtable;

import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.groupingBy;

public class JewelsAndStones {
    public int numJewelsInStones(String jewels, String stones) {
        var list = stones.chars().mapToObj(c -> (char) c).toList();
        Map<Character, Long> map = list.stream().collect(groupingBy(Function.identity(), Collectors.counting()));
        var result = 0L;
        for (char c : jewels.toCharArray()) {
            result += map.getOrDefault(c, 0L);
        }
        return (int) result;
    }

    public int numJewelsInStonesWithList(String jewels, String stones) {
        var jewelsList = jewels.chars().mapToObj(c -> (char) c).toList();
        var result = 0;
        for (int i = 0; i < stones.length(); i++) {
            result = jewelsList.contains(stones.charAt(i)) ? result + 1 : result;
        }
        return result;
    }
}

File: src/main/kotlin/ting/leetcode/hashtable/JewelsAndStones.kt
package ting.leetcode.hashtable

class JewelsAndStones {
    fun numJewelsInStones(jewels: String, stones: String): Int = jewels.sumOf { ch -> stones.count { it == ch } }

    fun numJewelsInStonesWithMap(jewels: String, stones: String): Int {
        val stonesMap = stones.withIndex().groupBy ({ it.value }, { it.index })
        var result = 0
        jewels.forEach { ch -> stonesMap[ch]?.let { result += it.size } }
        return result
    }
}